src/main/flight/gps_rescue.c

```diff
@@ -229,164 +218,238 @@ static void idleTasks()
+    // I component alone cannot exceed a pitch angle of 10%
 
     // D component
-    const int zVelocityDerivative = zVelocityError - previousZVelocityError;
-    previousZVelocityError = zVelocityError;
+    float velocityD = ((velocityError - previousVelocityError) / sampleIntervalNormaliseFactor);
+    previousVelocityError = velocityError;
+    // simple first order filter on derivative with k = 0.5 for 200ms steps
+    velocityD = previousVelocityD + rescueState.sensor.filterK * (velocityD - previousVelocityD);
+    previousVelocityD = velocityD;
+    velocityD *= gpsRescueConfig()->velD;
+
+    // Pitch PIDsum and smoothing
+    float pitchAdjustment = velocityP + velocityD + velocityI;
+    // simple rate of change limiter - not more than 25 deg/s of pitch change to keep pitch smooth
+    float pitchAdjustmentDelta = pitchAdjustment - previousPitchAdjustment;
+    if (pitchAdjustmentDelta > rescueState.sensor.maxPitchStep) {
+        pitchAdjustment = previousPitchAdjustment + rescueState.sensor.maxPitchStep;
+    } else if (pitchAdjustmentDelta < -rescueState.sensor.maxPitchStep) {
+        pitchAdjustment = previousPitchAdjustment - rescueState.sensor.maxPitchStep;
+    }
+    const float movingAvgPitchAdjustment = 0.5f * (previousPitchAdjustment + pitchAdjustment);
+     // moving average seems to work best here, a lot of sequential up and down in velocity data
+    previousPitchAdjustment = pitchAdjustment;
+    pitchAdjustment = movingAvgPitchAdjustment;
+    // pitchAdjustment is the absolute Pitch angle adjustment value in degrees * 100
+    // it gets added to the normal level mode Pitch adjustments in pid.c
 
-    const int16_t hoverAdjustment = (hoverThrottle - 1000) / ct;
-    altitudeAdjustment = constrain(altitudeAdjustment + (throttle.Kp * zVelocityError + throttle.Ki * zVelocityIntegral + throttle.Kd * zVelocityDerivative),
-                                    gpsRescueConfig()->throttleMin - 1000 - hoverAdjustment, gpsRescueConfig()->throttleMax - 1000 - hoverAdjustment);
+    gpsRescueAngle[AI_PITCH] = constrainf(pitchAdjustment, rescueState.intent.minAnglePitchDeg * 100.0f, rescueState.intent.maxAnglePitchDeg * 100.0f);
+    // this angle gets added to the normal pitch Angle Mode control values in pid.c - will be seen in pitch setpoint
 
-    rescueThrottle = constrain(1000 + altitudeAdjustment + hoverAdjustment, gpsRescueConfig()->throttleMin, gpsRescueConfig()->throttleMax);
+    DEBUG_SET(DEBUG_GPS_RESCUE_VELOCITY, 0, velocityP);
+    DEBUG_SET(DEBUG_GPS_RESCUE_VELOCITY, 1, velocityD);
+    DEBUG_SET(DEBUG_GPS_RESCUE_VELOCITY, 3, rescueState.intent.targetVelocityCmS);
+    DEBUG_SET(DEBUG_GPS_RESCUE_TRACKING, 1, rescueState.intent.targetVelocityCmS);
 
-    DEBUG_SET(DEBUG_RTH, 0, rescueThrottle);
-    DEBUG_SET(DEBUG_RTH, 1, gpsRescueAngle[AI_PITCH]);
-    DEBUG_SET(DEBUG_RTH, 2, altitudeAdjustment);
+    /**
+        Altitude (throttle) controller
```

Comment:
I suggest to use controller from #11342 here

Reply:
Indeed, we only need one controller.  I have no particular affection for this controller.  I know it works quite well in the dynamic environment that it is presented with.
Please fork this PR, and demonstrate the benefits of a better controller by replacing this altitude code with something else. Then, we can make some logs comparing the two controllers, and see which works best.  We may find that by working together we get a better result than working independently.  
Once we have an optimised altitude controller, then I don't mind where that code goes.  It will be called not by GPS Rescue, or altitude hold, but altitude limit and also loiter or position hold.  So we definitely want a single good controller.
To be 100% clear,  before we add any "new features" to the code base, such as Altitude Hold, we have to make the GPS Rescue code, an existing feature, work properly.  One aspect of that is having excellent altitude control.  I welcome your assistance with this.

Reply:
Ok